go/vt/vtgate/gateway/discoverygateway.go

```diff
@@ -273,13 +273,51 @@ func (dg *discoveryGateway) withRetry(ctx context.Context, target *querypb.Targe
 	return NewShardError(err, target, tabletLastUsed, inTransaction)
 }
 
-func shuffleTablets(tablets []discovery.TabletStats) {
-	index := 0
+func shuffleTablets(cell string, tablets []discovery.TabletStats) {
+	sameCell, diffCell, sameCellMax := 0, 0, -1
 	length := len(tablets)
-	for i := length - 1; i > 0; i-- {
-		index = rand.Intn(i + 1)
-		tablets[i], tablets[index] = tablets[index], tablets[i]
+
+	//move all same cell tablets to the front, this is O(n)
```

Comment:
Go thing: need space between `//` and start of comment. Running gofmt should fix this. Some pedantic people may also insist on punctuation.